src/lib/provable/bigint.ts

```diff
@@ -0,0 +1,898 @@
+        this.fields,
+        a.fields,
+        q.fields,
+        r.fields,
+        this.Constructor.modulus.fields,
+      ];
+
+      for (let i = 0; i < this.Constructor.config.limb_num; i++) {
+        if (isSquare) {
+          for (let j = 0; j < i; j++) {
+            delta[i + j] = delta[i + j].add(X[i].mul(X[j]).mul(2n));
+          }
+          delta[2 * i] = delta[2 * i].add(X[i].mul(X[i]));
+        } else {
+          for (let j = 0; j < this.Constructor.config.limb_num; j++) {
+            delta[i + j] = delta[i + j].add(X[i].mul(Y[j]));
+          }
+        }
+
+        for (let j = 0; j < this.Constructor.config.limb_num; j++) {
+          delta[i + j] = delta[i + j].sub(Q[i].mul(P[j]));
+        }
+
+        delta[i] = delta[i].sub(R[i]).seal();
+      }
+
+      let carry = new Field(0);
+
+      for (let i = 0; i < 2 * this.Constructor.config.limb_num - 2; i++) {
+        let deltaPlusCarry = delta[i].add(carry).seal();
+
+        carry = Provable.witness(Field, () =>
+          deltaPlusCarry.div(1n << BigInt(this.Constructor.config.limb_size))
+        );
+        rangeCheck(carry, 128, true);
+
+        deltaPlusCarry.assertEquals(carry.mul(1n << BigInt(this.Constructor.config.limb_size)));
+      }
+
+      delta[2 * this.Constructor.config.limb_num - 2].add(carry).assertEquals(0n);
+
+      return r;
+    }
+
+    /**
+     * Computes the square root of a ProvableBigInt
+     * @returns The square root as a ProvableBigInt
+     */
+    square(): ProvableBigInt_ {
+      return this.mul(this, true);
+    }
+
+    /**
+     * Divides one ProvableBigInt by another.
+     * @param a The divisor as a ProvableBigInt
+     * @returns The quotient as ProvableBigInt
+     */
+    div(a: ProvableBigInt_): ProvableBigInt_ {
+      const inv_a = a.inverse();
+
+      let res = this.mul(inv_a);
+
+      return res;
+    }
+
+    /**
+     * Computes the modular inverse of a ProvableBigInt
+     * @returns The inverse as a ProvableBigInt
+     */
+    inverse(): ProvableBigInt_ {
+      let { res } = Provable.witness(
+        Struct({ res: ProvableBigInt_ as typeof ProvableBigInt }),
+        () => {
+          const p = this.Constructor.modulus.toBigInt();
+          let t = 0n;
+          let newT = 1n;
+          let r = p;
+          let newR = this.toBigInt();
+
+          // Loop until newR is equal to zero
+          while (newR !== 0n) {
+            const quotient = r / newR;
+
+            [t, newT] = [newT, t - quotient * newT];
+            [r, newR] = [newR, r - quotient * newR];
+          }
+
+          // If r is bigger than 1, a is not invertible
+          if (r > 1n) {
+            throw new Error('a is not invertible');
+          }
+
+          // If t is smaller than zero, add modulus
+          if (t < 0n) {
+            t = t + p;
+          }
+
+          return { res: ProvableBigInt_.fromBigInt(t) };
+        }
+      );
+
+      res.mul(this).assertEquals(ProvableBigInt_.one());
```

Comment:
Nice, "expensive" witness computation block but then in circuit all we want is this assertion.